### PR TITLE
WifiDirect UI refactor

### PR DESCRIPTION
**Pull request by wpride**

### CODE REVIEW

**Comment by wpride** [comment]
new wifi direct

**Comment by wpride** [comment]
Using square buttons for Wifi Direct and some other minor fixes. Open for review, but I think I'll want to combine some of the adapter code with @phillipm home screen code in a base class

File: app/res/drawable/wifi_direct_change_mode.xml [hunk 1]
```diff
@@ -0,0 +1,5 @@
+<?xml version="1.0" encoding="UTF-8"?>
+<vector android:height="512dp" android:viewportHeight="24.0"
+    android:viewportWidth="24.0" android:width="512dp" xmlns:android="http://schemas.android.com/apk/res/android">
+    <path android:fillColor="#FF000000" android:pathData="M19.43,12.98c0.04,-0.32 0.07,-0.64 0.07,-0.98s-0.03,-0.66 -0.07,-0.98l2.11,-1.65c0.19,-0.15 0.24,-0.42 0.12,-0.64l-2,-3.46c-0.12,-0.22 -0.39,-0.3 -0.61,-0.22l-2.49,1c-0.52,-0.4 -1.08,-0.73 -1.69,-0.98l-0.38,-2.65C14.46,2.18 14.25,2 14,2h-4c-0.25,0 -0.46,0.18 -0.49,0.42l-0.38,2.65c-0.61,0.25 -1.17,0.59 -1.69,0.98l-2.49,-1c-0.23,-0.09 -0.49,0 -0.61,0.22l-2,3.46c-0.13,0.22 -0.07,0.49 0.12,0.64l2.11,1.65c-0.04,0.32 -0.07,0.65 -0.07,0.98s0.03,0.66 0.07,0.98l-2.11,1.65c-0.19,0.15 -0.24,0.42 -0.12,0.64l2,3.46c0.12,0.22 0.39,0.3 0.61,0.22l2.49,-1c0.52,0.4 1.08,0.73 1.69,0.98l0.38,2.65c0.03,0.24 0.24,0.42 0.49,0.42h4c0.25,0 0.46,-0.18 0.49,-0.42l0.38,-2.65c0.61,-0.25 1.17,-0.59 1.69,-0.98l2.49,1c0.23,0.09 0.49,0 0.61,-0.22l2,-3.46c0.12,-0.22 0.07,-0.49 -0.12,-0.64l-2.11,-1.65zM12,15.5c-1.93,0 -3.5,-1.57 -3.5,-3.5s1.57,-3.5 3.5,-3.5 3.5,1.57 3.5,3.5 -1.57,3.5 -3.5,3.5z"/>
+</vector>
```
Comment (phillipm) [inline app/res/drawable/wifi_direct_change_mode.xml]: Glad to see you got the vector graphics importer working, pretty cool stuff, though finicky at times. Can you please include the original svg file in the `app/original_vector_graphics/`dir?

File: app/src/org/commcare/android/adapters/WiFiDirectAdapter.java [hunk 2]
```diff
@@ -0,0 +1,184 @@
...
+    }
+
+    @Override
+    public void onBindViewHolder(RecyclerView.ViewHolder holder, int position) {
+        bindCard((SquareButtonViewHolder)holder, position);
+    }
+
+    private void bindCard(SquareButtonViewHolder holder, int position) {
+        HomeCardDisplayData cardDisplayData = getItem(position);
+        String notificationText = null;
```
Comment (phillipm) [inline app/src/org/commcare/android/adapters/WiFiDirectAdapter.java]: remove and replace single usage with `null`

**Comment by phillipm** [comment]
If you rotate the screen and don't specify a mode, then rotate again and choose a mode, the app crashes. I would modify the code to only show the mode selection dialog when you launch the activity and not on rotation. I usually check this condition by testing if the `Bundle` passed into `onCreate` is `null`.

**Comment by phillipm** [comment]
The buttons are hidden in landscape mode and on smaller screens.

**Comment by phillipm** [comment]
Since I initially wrote `HomeScreenAdapter` I wanted to generalize it. I pushed a commit to this branch lifting common logic between `HomeScreenAdapter` and `WiFiDirectAdapter` into `SquareButtonAdapter`; hope you're cool with that.

File: app/src/org/commcare/android/adapters/WiFiDirectAdapter.java [hunk 3]
```diff
@@ -0,0 +1,109 @@
...
+    }
+
+    @Override
+    public int getItemCount() {
+        switch (wiFiDirectActivity.mState) {
+            case send:
+                return 3;
+            case submit:
+                return 2;
+        }
```
Comment (phillipm) [inline app/src/org/commcare/android/adapters/WiFiDirectAdapter.java]: Switch is missing the `receive` state

File: app/src/org/commcare/android/adapters/WiFiDirectAdapter.java [hunk 4]
```diff
@@ -0,0 +1,109 @@
...
+        HomeCardDisplayData submitButton = HomeCardDisplayData.homeCardDataWithStaticText("Submit",
+                R.color.white,
+                R.drawable.wifi_direct_submit,
+                R.color.solid_dark_orange,
+                getSubmitButtonListener(wiFiDirectActivity));
+        HomeCardDisplayData changeModeButton = HomeCardDisplayData.homeCardDataWithStaticText("Change Mode",
+                R.color.white,
+                R.drawable.wifi_direct_change_mode,
+                R.color.cc_brand_text,
+                getChangeModeButtonListener(wiFiDirectActivity));
```
Comment (phillipm) [inline app/src/org/commcare/android/adapters/WiFiDirectAdapter.java]: Ideally we wouldn't create new `HomeCardDisplayData` objects every time `getItem` is called. A list of them can be stored statically and filtered every time `getButtonDisplayData` is called. There are probably even better ways of doing it using some sort of listener for changes to the `wiFiDirectActivity.mState`, but those might be overly complex.

**Comment by phillipm** [comment]
Done with a first code review pass.

File: app/src/org/commcare/android/util/FormUploadUtil.java [hunk 5]
```diff
@@ -293,7 +293,7 @@ private static boolean buildMultipartEntity(MultipartEntity entity,
                     fb = new EncryptedFileBody(f, FormUploadUtil.getDecryptCipher(key),
                             ContentType.TEXT_XML);
                 } else {
-                    fb = new FileBody(f, "text/xml");
+                    fb = new FileBody(f, ContentType.TEXT_XML, f.getName());
```
Comment (wpride) [inline app/src/org/commcare/android/util/FormUploadUtil.java]: @ctsims this was the fix for the submission issue - the deprecated format no longer passes HQ's newly strict requirements. Guessing we'll need/want to do the same for the other formats below

cc @benrudolph

File: app/src/org/commcare/android/adapters/WiFiDirectAdapter.java (hunk 3 above)
Comment (wpride) [inline app/src/org/commcare/android/adapters/WiFiDirectAdapter.java]: only one button in receive mode so base case is correct, adding comment

**Comment by wpride** [comment]
Release note: Wifi Direct UI refactor, bug fixes, localization strings, no longer require SD card

**Comment by wpride** [comment]
@phillipm thanks for the review. I opted to turn off landscape mode since it didn't feel worthwhile to build a separate awesome-mode-esque layout for this at the moment and there's a lot of information to display

File: app/src/org/commcare/android/util/FormUploadUtil.java (hunk 5 above)
Comment (benrudolph) [inline app/src/org/commcare/android/util/FormUploadUtil.java]: :+1:

File: app/src/org/commcare/dalvik/activities/CommCareWiFiDirectActivity.java [hunk 6]
```diff
@@ -175,9 +139,50 @@ public void onClick(View v) {
...
+                }
+
+                grid.requestLayout();
+                adapter.notifyDataSetChanged();
+                configUI();
+            }
+        });
+    }
+
+    private void configUI() {
```
Comment (phillipm) [inline app/src/org/commcare/dalvik/activities/CommCareWiFiDirectActivity.java]: Unneeded

File: app/src/org/commcare/dalvik/activities/CommCareWiFiDirectActivity.java [hunk 7]
```diff
@@ -175,9 +139,50 @@ public void onClick(View v) {
...
+        final RecyclerView grid = (RecyclerView)findViewById(R.id.wifi_direct_gridview_buttons);
+        grid.setHasFixedSize(true);
+
+        StaggeredGridLayoutManager gridView =
+                new StaggeredGridLayoutManager(3, StaggeredGridLayoutManager.VERTICAL);
+        grid.setLayoutManager(gridView);
+        grid.setItemAnimator(null);
+        grid.setAdapter(adapter);
+
+        grid.getViewTreeObserver().addOnGlobalLayoutListener(new ViewTreeObserver.OnGlobalLayoutListener() {
```
Comment (phillipm) [inline app/src/org/commcare/dalvik/activities/CommCareWiFiDirectActivity.java]: Not positive, but I don't think this listener is needed. @ctsims any reason this should be here? And can you give context as to why it is needed on the home screen adapter? Can it be removed there too?

File: app/src/org/commcare/dalvik/activities/CommCareWiFiDirectActivity.java [hunk 8]
```diff
@@ -17,13 +18,21 @@
...
 import android.view.View.OnClickListener;
+import android.view.ViewGroup;
+import android.view.ViewTreeObserver;
+import android.widget.BaseAdapter;
 import android.widget.Button;
 import android.widget.TextView;
 import android.widget.Toast;
 
+import org.commcare.android.adapters.HomeScreenAdapter;
+import org.commcare.android.adapters.SquareButtonViewHolder;
```
Comment (phillipm) [inline app/src/org/commcare/dalvik/activities/CommCareWiFiDirectActivity.java]: There are some unused imports here

**Comment by phillipm** [comment]
Good to merge after my comments are addressed

**Comment by wpride** [comment]
good to merge or wait on @ctsims to weigh in on TreeObserver

**Pull request merged**